clients/drcachesim/reader/v2p_reader.cpp

```diff
@@ -33,83 +33,114 @@
+    const std::string page_size_key = "page_size";
+    const std::string page_count_key = "page_count";
+    const std::string bytes_mapped_key = "bytes_mapped";
+    const std::string virtual_address_key = "virtual_address";
+    const std::string physical_address_key = "physical_address";
+    // Assumes virtual_address 0 is not in the v2p file.
+    addr_t virtual_address = 0;
     std::string line;
     while (std::getline(file, line)) {
-        std::size_t found = line.rfind(virtual_address_key);
+        std::size_t found = line.find(virtual_address_key);
         if (found != std::string::npos) {
-            std::stringstream ss(line);
-            std::string elem;
-            if (!std::getline(ss, elem, ':')) {
-                ERRMSG("ERROR: virtual_address key not found. Found %s instead.\n",
-                       elem.c_str());
-                return false;
+            std::vector<std::string> key_val_pair = split_by(line, separator);
+            if (key_val_pair.size() != 2) {
+                error_ss << "ERROR: " << virtual_address_key
+                         << " key or value not found.";
+                return error_ss.str();
             }
-            if (!std::getline(ss, elem, ':')) {
-                ERRMSG("ERROR: virtual_address value not found. Found %s instead.\n",
-                       elem.c_str());
-                return false;
-            }
-            virtual_address = std::stoull(elem, nullptr, 0);
+            virtual_address =
+                static_cast<addr_t>(std::stoull(key_val_pair[1], nullptr, 0));
             continue;
         }
-        found = line.rfind(physical_address_key);
+
+        found = line.find(physical_address_key);
         if (found != std::string::npos) {
+            std::vector<std::string> key_val_pair = split_by(line, separator);
+            if (key_val_pair.size() != 2) {
+                error_ss << "ERROR: " << physical_address_key
+                         << " key or value not found.";
+                return error_ss.str();
+            }
             if (virtual_address == 0) {
-                ERRMSG("ERROR: no corresponding virtual_address for this "
-                       "physical_address.\n");
-                return false;
+                error_ss << "ERROR: no corresponding " << virtual_address_key
+                         << " for this " << physical_address_key << " " << key_val_pair[1]
+                         << ".";
+                return error_ss.str();
             }
-            std::stringstream ss(line);
-            std::string elem;
-            if (!std::getline(ss, elem, ':')) {
-                ERRMSG("ERROR: physical_address key not found. Found %s instead.\n",
-                       elem.c_str());
-                return false;
+            addr_t physical_address =
+                static_cast<addr_t>(std::stoull(key_val_pair[1], nullptr, 0));
+            v2p_info.v2p_map[virtual_address] = physical_address;
```

Comment:
Should we check whether an entry already exists and consider that an error?

Reply:
Yes, done.